### DISCUSSION TITLE
GDL on OSX 11.1 Big Sur

### DISCUSSION THREAD

**Discussion by alaingdl** [gnudatalanguage/gdl#889] [General]
Following first report on GDL on OSX 11.1 here https://github.com/gnudatalanguage/gdl/issues/873#issuecomment-765679886 
I made some progress but I had also a lot of problems : 

* minimal_script.sh is not working at the GDL stage, due to an issue with readline
* with GCC 10 (from Brew), with and without OpenMP, I cannot compile.
* with LLVM/Clang 10, I cannot activate OpenMP
* I do have a serious issue with readline, then I need to desactivate it
* (up to now) no brew package for : hdf (hdf5 and netcdf OK), udunits2

What can be useful : X11 headers from Brew are in libx11

At the end, I succeed with : 
```
$ cmake .. -DREADLINE=no -DMAGICK=OFF -DWXWIDGETS=OFF  -DHDF=OFF -DPYTHON=OFF -DUDUNITS2=OFF  -DOPENMP=off
```

Please note that I am working remotely on that machine, and up to now I did not succeed to switch on the X11 forwarding. (X11 forwarding request failed on channel 0). I do have solutions for all previous OSX versions from 10.6 to 10.14 ... (why do they change tricky details in those /etc/ssh files ?!)

**Comment by kkin1995** [comment]
Hi, I just installed GDL from source on macOS Big Sur. This was the process I followed:

1. Install command line tools 

```xcode-select --install```

2. Clone the GDL repository to your local machine 

```git clone https://github.com/gnudatalanguage/gdl```

3. Set the working directory to the GDL directory 

```cd gdl```

4. Make a directory to store all the built files

```mkdir build```

5. ```cd build```

6. 
```
cmake .. -DREADLINE=no -DPNGLIB=OFF -DOPENMP=OFF
-DGSLDIR=/usr/local/Cellar/gsl/2.6
-DPLPLOTDIR=/usr/local/Cellar/plplot/5.15.0_2 \
-DMAGICK=OFF -DWXWIDGETS=OFF -DTIFF=OFF -DNETCDF=OFF
-DHDF=OFF -DHDF5=OFF -DFFTW=OFF -DLIBPROJ4=OFF
-DPYTHON=off -DUDUNITS2=OFF -DEIGEN3=OFF -DGRIB=OFF
-DGLPK=OFF -DSHAPELIB=OFF -DX11=OFF
```

Note: If the 6th command says that some packages are not installed, they need to be installed (can be done through homebrew) and then re-run the 6th command.

7. ```make -j 4```

8. ```make check```

Note: Some tests in command 8 may fail

9. Create a "gdl-startup.pro" file in the gdl directory that was created and add the following line to the file:

```!PATH=/Users/<youraccount>/gdl/src/pro/```

Change <youraccount> to the name of the home folder on your Mac

Then, add the following line to your .profile file:

```export GDL_STARTUP='/Users/<youraccount>/gdl/gdl-startup.pro'```

10. To run gdl from the PATH, an alias has to be added to the profile. For Example, if you are using a ZSH terminal, the profile is found at ~/.zshrc

Add the following line to your .profile,

``` alias gdl='/Users/<youraccount>/gdl/build/src/gdl'```

Change <youraccount> to the name of the home folder on your mac

**Comment by alaingdl** [comment]
Thanks. We still have several important problems on Big Sur : readline, openmp are critical.
I think it will be simple and OK for Eigen3.

**Reply by kkin1995** [reply]
Modified command 6 to include Readline and X11 Window System (since I was experiencing problems with the graphic device):

```
cmake .. -DREADLINE=/usr/local/Cellar/readline/8.1 -DPNGLIB=OFF -DOPENMP=OFF
-DGSLDIR=/usr/local/Cellar/gsl/2.6
-DPLPLOTDIR=/usr/local/Cellar/plplot/5.15.0_2 \
-DMAGICK=OFF -DWXWIDGETS=OFF -DTIFF=OFF -DNETCDF=OFF
-DHDF=OFF -DHDF5=OFF -DFFTW=OFF -DLIBPROJ4=OFF
-DPYTHON=off -DUDUNITS2=OFF -DEIGEN3=OFF -DGRIB=OFF
-DGLPK=OFF -DSHAPELIB=OFF -DX11=/usr/local/Cellar/libx11/1.7.0
```

However, I still was experiencing issues and got the error:
```
PLplot installation lacks the requested driver: xwin
```

**Comment by GillesDuvert** [comment]
Is wxWidgets available? In the future releases of GDL wxWidgets will be preferred wrt X11.

**Reply by kkin1995** [reply]
Theres a homebrew package called wxmac (also called wxwidgets) available.

**Comment by alaingdl** [comment]
Thanks. At that time, I have Big Sur on a remote server and the ssh -X is not OK 
(X11 forwarding request failed on channel 0) and this time I did not succeed to locate "xauth" ...

**Comment by alaingdl** [comment]
I succeed to compiled with
```
 cmake .. -DREADLINEDIR=/usr/local/Cellar/readline/8.1/ -DMAGICK=OFF   -DHDF=OFF -DPYTHON=OFF -DOPENMP=off
```
after installing along Brew : fftw, wxmac, readline, gsl, plplot, udunits, glpk, shapelib, hdf5, netcdf, eccodes

but when launching GDL, I got 2 problems : 
1/ GDL> plot, findgen(10)
% PLplot installation lacks the requested driver: xwin

2/ large amount of messages (I suspect related to widgets) like that 

./src/common/object.cpp(251): assert "classTable->Get(m_className) == NULL" failed in Register(): Class "wxTextEntryDialog" already in RTTI table - have you used IMPLEMENT_DYNAMIC_CLASS() multiple times or linked some object file twice)?
./src/common/object.cpp(251): assert "classTable->Get(m_className) == NULL" failed in Register(): Class "wxPasswordEntryDialog" already in RTTI table - have you used IMPLEMENT_DYNAMIC_CLASS() multiple times or linked some object file twice)?
./src/common/object.cpp(251): assert "classTable->Get(m_className) == NULL" failed in Register(): Class "wxToolbook" already in RTTI table - have you used IMPLEMENT_DYNAMIC_CLASS() multiple times or linked some object file twice)?
.

**Reply by GillesDuvert** [reply]
with the trunk gdl, or the current fork aiming at unifying graphic windows and widgets on all platforms with wxWidget?

**Reply by GillesDuvert** [reply]
wxTextEntryDialog, wxPasswordEntryDialog and wxToolbook are not in GDL. wxmac is bugged?
xwin: please look at GraphicsDevice::Init() logic (#ifdef) that may be wrong on a mac.

**Comment by alaingdl** [comment]
ha, yes sorry you are right, I will try ASAP with your fork. BUT I will be able to have outputs since my remote display with SSH is not working

**Reply by GillesDuvert** [reply]
works with ssh -X on my machine, why not for you?

**Comment by alaingdl** [comment]
fatal: unable to access 'https://github.com/Homebrew/brew/': Failed to connect to github.com port 443: Operation timed out
this stupid issue and solution up to now ...

**Reply by alaingdl** [reply]
another bad news on the OSX side :

* script _minimum__script4gdl.sh_ works fine on 10.14.4 but fails on 11.1
(same version for cmake !)

* xwin drivers seems to be NOT present in Brew (but is OK when plplot compiled with  _minimum__script4gdl.sh_ )

**Comment by pjb7687** [comment]
I solved the OpenMP problem by adding
```
LIBRARY_PATH=$LIBRARY_PATH:/usr/local/opt/llvm/lib
```
before running CMake.

FYI, see this commit which includes the changes that I have made to the CI script (.travis.yml): https://github.com/gnudatalanguage/gdl/pull/866/commits/5ae085fe83037548af57b6906f10374e21294dc3

This also includes enabling libgeotiff and libx11. The build seems to work fine (https://travis-ci.com/github/gnudatalanguage/gdl/builds/216984505).

**Comment by GillesDuvert** [comment]
For what is worth, this is not on BigSur but Catalina, that I've installed in a VM, starting from scratch, installing homebrew and all the dependencies sucesfully as depicted in Travis.yml :
I **succeeded quite easily to have a full fledged version of gdl**.
One sole exception, was to recompile wxmac after brew install (see PR #882  file scripts/brew_enable_wxwidgets) to get the wxwidgets driver that is not 'made' by brew. OK, it does not work on Travis, but it works on my pristine Catalina.

So the travis.yml script is THE working full script, even if it 'looks' too travis-centered. I am as you have seen trying to simplify the GDL maintainance. There is IMHO need to replace the so-called minimal script of @alaingdl by the full-fledged 'travis' script, and have travis execute this full-fledged script, as we would maintain one and only one way to build GDL on mac.
If perchance one could persuade the script to make an installer/disk image then we could just tell people to get the latest built artifact., same as with the installer for windows.

What do you think?

**Reply by pjb7687** [reply]
Actually, I also experienced the full build didn't work on Travis but works pretty well on Github Actions. Here is my experimental attempt to build on Github Actions: https://github.com/gnudatalanguage/gdl/actions/runs/566393447

I would say we can also use Github Actions to build Linux and macOS builds then, based on the 'minimal' script. Would that be okay? I will try this during this weekend...

**Reply by slayoo** [reply]
I vote for phasing out Travis in favour of Github Actions

**Reply by GillesDuvert** [reply]
My mistake, I should have added that I favor GH actions, that in all 'MINIMAL' scripts  one must add -DINTERACTIVE_GRAPHICS=NO  otherwise it is not going to compile.

I personnaly doubt the utility of the 'minimal' script, as it tests nothing.
